govcd/edgegateway.go

```diff
@@ -311,87 +336,172 @@ func (eGW *EdgeGateway) AddSNATRule(networkHref, externalIP, internalIP, descrip
 	}
 
-	return eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule, nil
+	for _, natRule := range eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule {
+		if natRule.Description == mappingId {
+			createdNatRule = natRule
+			break
+		}
+	}
+
+	if createdNatRule == nil {
+		return nil, fmt.Errorf("error creating SNAT rule, didn't match created rule")
+	}
+
+	createdNatRule.Description = description
+
+	return eGW.UpdateNatRule(createdNatRule)
+}
+
+// getPseudoUuid creates unique ID/UUID
+func getPseudoUuid() (string, error) {
+
+	b := make([]byte, 16)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+
+	uuid := fmt.Sprintf("%X-%X-%X-%X-%X", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
+
+	return uuid, nil
+}
+
+// UpdateNatRule updates NAT rule and handles task. Returns updated NAT rule or error.
+func (eGW *EdgeGateway) UpdateNatRule(natRule *types.NatRule) (*types.NatRule, error) {
+	task, err := eGW.UpdateNatRuleAsync(natRule)
+	if err != nil {
+		return nil, fmt.Errorf("error updating NAT rule: %#v", err)
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	return eGW.FetchNatRule(natRule.ID)
+}
+
+// UpdateNatRuleAsync updates NAT rule and returns task or error.
+func (eGW *EdgeGateway) UpdateNatRuleAsync(natRule *types.NatRule) (Task, error) {
+	if "" != natRule.GatewayNatRule.Protocol && !isValidProtocol(natRule.GatewayNatRule.Protocol) {
+		return Task{}, fmt.Errorf("provided protocol is not one of TCP, UDP, TCPUDP, ICMP, ANY")
+	}
+
+	if strings.ToUpper(natRule.GatewayNatRule.Protocol) == "ICMP" && !isValidIcmpSubType(natRule.GatewayNatRule.IcmpSubType) {
+		return Task{}, fmt.Errorf("provided icmp sub type is not correct")
+	}
+
+	err := eGW.Refresh()
+	if err != nil {
+		return Task{}, fmt.Errorf("error refreshing edge gateway: %#v", err)
+	}
+
+	natServiceToUpdate := eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService
+
+	if natServiceToUpdate != nil {
+		for n, existingNatRule := range natServiceToUpdate.NatRule {
+			if existingNatRule.ID == natRule.ID {
+				natServiceToUpdate.NatRule[n] = natRule
+			}
+		}
+	} else {
+		return Task{}, fmt.Errorf("edge gateway doesn't have such nat rule")
+	}
+
+	newRules := &types.EdgeGatewayServiceConfiguration{
+		Xmlns:      types.XMLNamespaceVCloud,
+		NatService: natServiceToUpdate,
+	}
+
+	egwConfigureHref, _ := url.ParseRequestURI(eGW.EdgeGateway.HREF)
+	egwConfigureHref.Path += "/action/configureServices"
+
+	// Return the task
+	return eGW.client.ExecuteTaskRequest(egwConfigureHref.String(), http.MethodPost,
+		"application/vnd.vmware.admin.edgeGatewayServiceConfiguration+xml", "error reconfiguring Edge Gateway: %s", newRules)
+}
+
+// FetchNatRule returns NAT rule or error.
+func (eGW *EdgeGateway) FetchNatRule(id string) (*types.NatRule, error) {
```

Comment:
I believe we agreed to use `Read...` after all. @dataclouder please confirm if nothing changed.

Reply:
I wasn't informed. You did meeting without me. I was against READ

Reply:
We agreed on Read.
See https://github.com/vmware/go-vcloud-director/issues/211

Reply:
Changed

Reply:
Re-met, re-evaluated & decided to go with `Get` everywhere.

Reply:
already changed